### PR TITLE
Improve and fix minimal tearing.

### PR DESCRIPTION
**Pull request by mahge**
- Minimal tearing now handles algorithm equations.

  - It now uses the non-scalarized adjacency list. This avoids
    having to re-process and construct complex equations in the end.

  - It now uses a local subsystem to do the matching. This is
      not strictly needed but simplifies things.

- [ ] TO DO

  - The inner equations of the torn system are NOT properly sorted right now.
    One way to do this is to generate the adjacency list again but not limiting
    it to solvable variables.
    Another way is to use the enhanced adjacency list and modify the matching
    accordingly.
  - The matching needs to be cleaned up a bit. It works as it is but
    can use some cleaning up.

### CODE REVIEW

File: OMCompiler/Compiler/BackEnd/Tearing.mo [hunk 1]
```diff
@@ -1648,77 +1649,88 @@ protected function minimalTearing
...
+  Integer size, i;
+  array<Integer> nE, nV, scalar_to_multi_dim_eq_mapping;
+  array<list<Integer>> collected_inner_eq_vars;
   array<Boolean> varArray, eqArray;
-  list<Integer> discreteVars={};
-  list<Integer> tearingvars = {}, residualequations = {};
-  list<Integer> tSel_always, tSel_prefer, tSel_avoid, tSel_never;
-  list<BackendDAE.Var> var_lst;
-  BackendDAE.InnerEquations helpInnerEquations = {};
+  list<Integer> discreteVars = {};
```
Comment (JKRT) [inline OMCompiler/Compiler/BackEnd/Tearing.mo]: Maybe initialize all lists to the empty list for uniformity

File: OMCompiler/Compiler/BackEnd/Tearing.mo [hunk 2]
```diff
@@ -1462,6 +1463,19 @@ algorithm
   end match;
 end solvable;
 
+protected function isEntrySolved
+  input BackendDAE.AdjacencyMatrixElementEnhancedEntry ent;
+  output Boolean b;
+algorithm
+  b := match ent
+    case (_, BackendDAE.SOLVABILITY_SOLVED(), _) then true;
+    case (_, BackendDAE.SOLVABILITY_PARAMETER(b=b), _) algorithm
```
[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: Here I think we are only interested in SOLVED variables. Not SOLVABLE ones. That is we are looking for: the algorithm outputs (assigned vars). I have put an error for the SOLVABILITY_PARAMETER because I did not understand what it was and wanted to see something fail for it to investigate further.

The concept of non-assigned variables of an algorithm **being marked as solvable** is what makes the non-enhanced adjacency matrix problematic. 

Non assigned variables in an algorithm section should not be considered solvable if we are not actually inverting the whole algorithm and solving it as a non-linear system.  We can't really rearrange one statement in a an algorithm section to solve for a specific variable.

**Comment by mahge** [comment]
> 
> 
> Looks good to me.
> It's way more clear now what's happening.
> 
> The remaining two failing tests are looking correct, only the order of variables changed. Maybe there was a `ListReverse` added somewhere?
> If they are running we should merge this

@AnHeuermann This is my own doing actually. Previous changes to the non-enhanced adjacency matrix changed some stuff and I updated the tests. I just need to revert them back.

File: testsuite/simulation/modelica/tearing/Tearing18-minimal.mos [hunk 3]
```diff
@@ -28,6 +28,9 @@ val(x8,1.0); getErrorString();
 //     resultFile = "Tearing18_res.mat",
 //     simulationOptions = "startTime = 0.0, stopTime = 1.0, numberOfIntervals = 500, tolerance = 1e-06, method = 'dassl', fileNamePrefix = 'Tearing18', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = ''",
 //     messages = "LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
+// assert | debug   | Solving non-linear system 34 failed at time=0.086.
```
Comment (AnHeuermann) [inline testsuite/simulation/modelica/tearing/Tearing18-minimal.mos]: Why does this test case fail at an assert (only once)?
The loop itself looks correct (nothing teared), so the behaviour should be the same as if no tearing at all was performed. The problem is that the non-linear loop itself is really hard to solve, so sometimes it runs and sometimes it fails with marginal different settings.

My suggestion is to use `status: should fail` (or what ever the equivalent is in out testsuite)

File: testsuite/simulation/modelica/tearing/userDefinedTearing2.mos [hunk 4]
```diff
@@ -67,14 +66,12 @@ val(c,0.5); getErrorString();
 //  * Record equations: 0
 //  * When equations: 0
 //  * If-equations: 0
-//  * Equation systems (linear and non-linear blocks): 1
-//  * Torn equation systems: 0
+//  * Equation systems (linear and non-linear blocks): 0
```
Comment (AnHeuermann) [inline testsuite/simulation/modelica/tearing/userDefinedTearing2.mos]: Does this work now because the adjacency matrix changed?

File: testsuite/simulation/modelica/tearing/Tearing18-minimal.mos (hunk 3 above)
Comment (mahge) [reply testsuite/simulation/modelica/tearing/Tearing18-minimal.mos]: Ya. Good catch. I was wondering why as well first. But then I realized the model behaves like this for all tearing methods. 
The test cases `Tearing18-cel.mos`, T`earing18-celMC3.mos`, `Tearing18-minimal.mos `and `Tearing18-omc.mos` behave the same for this non-linear system.

Comment (mahge) [reply testsuite/simulation/modelica/tearing/Tearing18-minimal.mos]: Actually on second look, for all the the other tearing methods (`Tearing18-cel.mos`, `Tearing18-celMC3.mos`, and `Tearing18-omc.mos` ) the simulation gets terminated by a assert. For the minimal one it seems to continue and finish properly.

Comment (AnHeuermann) [reply testsuite/simulation/modelica/tearing/Tearing18-minimal.mos]: Yes, but for `tearingMethod=noTearing` it will fail, too. But the system is exactly the same, so why is it behaving different?

Comment (mahge) [reply testsuite/simulation/modelica/tearing/Tearing18-minimal.mos]: I am not sure. As you said the NLS seems to be difficult so solve at that specific time. Any attempt of tearing seems to cause it to fail. Fail as in siumation gets terminated by assert. While the non-teared version (minimal and noTearing) seem to at least be able to continue and finish simulation.

File: OMCompiler/Compiler/BackEnd/Tearing.mo [hunk 5]
```diff
@@ -2241,6 +2321,38 @@ algorithm
   end for;
 end findDiscrete;
 
+protected function findDiscreteWarnTearingSelect
+"mahge: Finds discrete variables in a variable list. This will also warn if the
+discrete variable has tearing select annotations. Right now this is used in manadatory
+tearing. It should probably used everywhere else as long as we don't have mixed solvers
+since we can not chose a discrete variable as a tearing(iteration) variable anyway.
+Note: This function returns indices based on the order of the variables in the list."
```
Comment (kabdelhak) [inline OMCompiler/Compiler/BackEnd/Tearing.mo]: Strongly approve this one

File: OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo [hunk 6]
```diff
@@ -6113,7 +5854,7 @@ algorithm
       tuple<BackendDAE.Variables, Boolean, Boolean, Integer, array<Integer>, list<Integer>> tpl;
 
     case (DAE.LUNARY(exp=e1), (vars, bs, isInitial, it, at, pa)) equation
-      (_, (vars, _, _, _, _, pa)) = Expression.traverseExpTopDown(e1, traversingAdjacencyRowExpSolvableEnhancedFinder, (vars, true, isInitial, it, at, pa));
+      (_, (vars, _, _, _, _, pa)) = Expression.traverseExpTopDown(e1, traversingAdjacencyRowExpSolvableEnhancedFinder, (vars, bs, isInitial, it, at, pa));
```
Comment (kabdelhak) [inline OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo]: Somehow this did not make it in the last review so here is just a single comment:

Why did you change `true` to `bs`? What does `bs` stand for? Do we need to change this in all function calls below?

File: OMCompiler/Compiler/BackEnd/Tearing.mo (hunk 2 above)
[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: Yes right now it is used for algorithms only. Because we mark non-output (non-assigned) vars in algorithm statements as SOLVABLE (but not SOLVED). The idea of such a var being considered solvable (individually without treating the whole algorithm as its own NLS) is open for discussion :)

But in any case the function is used to filter out SOLVED variables. It is named `isEntrySolved()`. There is another one called `isEntrySolvable()` that I used in other places. It depends which one you use based on what you need I think.

File: OMCompiler/Compiler/BackEnd/Tearing.mo [hunk 7]
```diff
@@ -1648,100 +1671,163 @@ protected function minimalTearing
...
+    // in sumulation time.
+    // What is odd is that the order in adj matrix is not really guided
+    // by anything. Somehow this seems to be faster for some models.
+    // However, we now use the enahnced matrix which can sometimes have
+    // the var entries for an equation slightly different order than the normal
+    // adj matix.
+    // Simulation should not be affected by what order we have here but it is.
+    // There is something in later phases that operates according to this order
+    // But it should not. This is just random order.
+    /*
```
[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: I just had a discussion with @lochel about this. And his suggestion is that it can vary from model to model. What bothered me was the amount of variation seems to be so big. 

It seems there is something we have to further investigate on how the Jacobian handling creates the sparsity and color structure for the systems. It is one of the big differences I saw in the generated code.

The good news is that if you use **kinsol** as your solver then the performance difference disappears completely for all orderings.

It is a bit involved to explain here but I will try. So we are considering two ways of ordering for the iteration vars in this seemingly rudimentary integer list. Say we have three residual equations, 1,2 and 3. Eq 1 uses vars 2 and 3. Eq 2 uses vars 1 and 2. 
1 -> {2,3}
2 -> {1,2}
3 -> {1,2,3}

The original ordering we were creating for the iteration vars was (let's call this with LOCALITY)
2,3,1

because we go through the residual equations in order and add the variables they used. I thought this was ineffective so I changed it to just iterate over the variables and add the as iteration if they are not matched. So we get the order (let's call this INCREASING ORDER)
1,2,3

What is surprising is that if the matrix looked like this (let's say if we use the enhanced matrix)
1 -> {3,2}**
2 -> {1,2}
3 -> {1,2,3}

The LOCALITY ordering will be 
3,2,1

and this seems to have an effect on simulation performance. 


------------------------------------------------------------------
Here is a full example for on NLS of a [PowerGrids Model](https://trac.openmodelica.org/OpenModelica/ticket/5797)
------------------------------------------------------------
Non-Enhanced LOCALITY ORDERING
* tVars: 42,38,41,20,40,21,39,37,24,22,19,36,35,34,33,30,2,32,17,31,4,18,29,28,27,26,25,23,1,16,3,15,14,6,5

* resEq: 36,35,34,33,32,31,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1

**sim time - 1.954s
sim-kinsol - 0.749s**

Enhanced LOCALITY ORDERING
* tVars: 38,42,41,20,40,21,24,39,22,37,19,36,35,34,30,33,2,32,17,31,4,18,29,28,27,26,25,23,1,16,3,15,14,6,5

* resEq: 36,35,34,33,32,31,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1

**sim - 7.297s
sim-kinsol - 0.733s**

Enhanced INCREASING ORDER
* tVars: 1,2,3,4,5,6,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42

* resEq: 36,35,34,33,32,31,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1

**sim - 14.834s
sim-kinsol - 0.746s**


As you can see there is quite a big differnce in simulation time. What is even more intersting is that oredring of variables matters even for one single equation. E.g.  Residual equation 36 uses vars 42 and 38. But the normal and enahanced adj matrices order them differently as you said, {42,38} and {38,42} respectively. This changes in ordering take the simulation time from 1.9 sec to 7.297 secs.

I do not know enough to fully interpret what is happening but I though it should be at least highlighted.

File: OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo (hunk 6 above)
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo]: Ya I know :( I wanted to change it but it will clutter things in this commit. 

The variable signifies "unsolvable". If it is true it means the exp is unsolvable. So as we traverse down we will mark all vars we find as unsolvable. 

It was set to true here which meant it was marking the vars in the exp of the LUNARY() all unsolvable. 

I would like to change it to at least say "solvable" and switch the boolean values around.  The best way would be to create a record instead of passing a tuple and match is with names like "bs".

File: OMCompiler/Compiler/BackEnd/Tearing.mo (hunk 7 above)
[the remark this answers is not in the document]
Comment (kabdelhak) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: For sure, this is quite frightening... I guess these are bad quirks of dense solvers. Maybe we should talk about this in more detail when we have some time!

File: OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo (hunk 6 above)
Comment (kabdelhak) [reply OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo]: Ah i see, so when i enter let's say a function call dependency list, i mark all occurring variables with unsolvable.

File: OMCompiler/Compiler/BackEnd/Tearing.mo (hunk 7 above)
[the remark this answers is not in the document]
Comment (casella) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: In the case of this model, there is clearly no question that a sparse solver should be used, and there is absolutly no point in using a dense solver and trying to optimize it. Would you care about optimizing an oxen-powered carriage when you have a diesel truck at hand? 😄 

In fact, not only a dense solver should not be used, but we shouldn't even use odeMode, and rather use daeMode in this case, because only daeMode keeps the structure of the model sparse until the end.

What we do need it to set the defaults for the selection solvers and solver mode (ode vs. dae) so that the right one is selected automatically. First of all, if one really wanted to use odeMode, a sparse nonlinear solver should be used (currently we only have kinsol, which is ok). I guess the current default limits of size and density for this selection are too conservative in favour of dense solvers. 

Second, we should probably suggest to try daeMode instead. Dymola already does it. As soon as daeMode is good enough with events, see [#5860](https://trac.openmodelica.org/OpenModelica/ticket/5860)

[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: @casella I agree completely on the issue of choosing the proper solver. 

The reason I added those comments to the code was that we need to figure out (just in case) why the order of variable detection in a single equation has that much performance effect for dense solvers. This issue will be there for any tearing method, not just minimalTearing. If we change up how we traverse a simple expression `a+b` vs` b+a` there is a significant difference in performance. The dense solver seems to be significantly affected by this order. I assume that some other heuristic that deals with non-linear systems after tearing needs some revising. Because it seems to be noticeably inconsistent for very small differences.

File: OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo (hunk 6 above)
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/BackendDAEUtil.mo]: Yes. I think so.

File: OMCompiler/Compiler/BackEnd/Tearing.mo (hunk 7 above)
[the remark this answers is not in the document]
Comment (casella) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: > 
> 
> @casella I agree completely on the issue of choosing the proper solver.
> 
> The reason I added those comments to the code was that we need to figure out (just in case) why the order of variable detection in a single equation has that much performance effect for dense solvers. This issue will be there for any tearing method, not just minimalTearing. If we change up how we traverse a simple expression `a+b` vs` b+a` there is a significant difference in performance. The dense solver seems to be significantly affected by this order. I assume that some other heuristic that deals with non-linear systems after tearing needs some revising. Because it seems to be noticeably inconsistent for very small differences.

I agree. However, I'm not sure I fully understand what is the difference in the three cases:
- are the tearing and torn variables the same?
- is the coloring of the Jacobian the same?
- is the variability in the ordering of torn equations made possible by the fact that only a partial ordering is enforced by the dependency graph?

[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: 1. The torn/iteration variables chosen are the same.
2. The coloring of the jacobian seems to be different. At least from what I can see in the generated code.
3. Yes? If I interpreted this question correctly. The end result in any case is that the ordering in the list of (integer indices) of variables chosen as iteration variables affects the simulation performance considerably.

If you change the ordering, of course, you expect some change. I understand that but my assumption was that any processing that depends on such ordering (e.g. sparsity) should be improved to decide the detection and ordering itself. Because at tearing time, we can't really make an informed decision about this. a+b+c might be collected as {1,2,3} or {2,3,1} depending on the traversal functions we use (which are different for the two types of matrices we have). And this is what caused the difference in simulation time. 

Anyway this is just one model. And it might very well be something we can not avoid or is the expected behavior.

[the remark this answers is not in the document]
Comment (casella) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: Wait, I find it a bit odd that the Jacobian colouring is influenced by the ordering of the torn equations and variables.

I lack some basic theoretical understanding here, please help me with that. Should the colouring of the Jacobian be unique, or can there be legitimate different choices for that? In the latter case, I guess we should consider heuristics to make sure that as many colours as possible are generated. Was that ever taken into consideration?

[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: > I lack some basic theoretical understanding here
Imagine me then :). 

> please help me with that. Should the colouring of the Jacobian be unique, or can there be legitimate different choices for that?
I have no idea really.  i do not know much how the sparsity pattern or coloring is actually used. I was hoping you would know what was going on or should be expected actually.


> Wait, I find it a bit odd that the Jacobian colouring is influenced by the ordering of the torn equations and variables.
Yup. That is where I was curious as well. To make it worse, it is not just the ordering of the equations, but the ordering of the detection of used variables for each equation (which affects the final order)

> I guess we should consider heuristics to make sure that as many colours as possible are generated. Was that ever taken into consideration?
As far as I can see the "number of colors"? was different for the two orderings.

[the remark this answers is not in the document]
Comment (casella) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: @AnHeuermann, @kabdelhak , I guess we need your comment here 😃

[the remark this answers is not in the document]
Comment (kabdelhak) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: Taken from [What Color Is Your Jacobian?](https://pdfs.semanticscholar.org/ae4d/65769b6551a51d6fc6be2f021515bffa0798.pdf):

11.3. Bounds for the Chromatic Number. Since computing the chromatic number χ(G) of a general graph G is NP-hard, we consider bounding it with more tractable graph parameters.

The problem of (perfect) coloring is NP-Hard so it is computed by a heuristic which can change depending on the order of equations and variables.

[the remark this answers is not in the document]
Comment (kabdelhak) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: I don't know if there is a bug or if it is just that sensitive.

[the remark this answers is not in the document]
Comment (mahge) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: On a slightly unrelated note 
[What Color Is Your Jacobian?](https://pdfs.semanticscholar.org/ae4d/65769b6551a51d6fc6be2f021515bffa0798.pdf): 
is actually written by my cousin. I guess I should read it from cover to cover.  😊😊

[the remark this answers is not in the document]
Comment (casella) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: > I don't know if there is a bug or if it is just that sensitive.

I'd say that the problem that raised @mahge attention is definitly overstretched to be handled by tearing and dense solvers. The question is whether or not we can get similar differences in performance (1:10) in problems where tearing and dense solvers are appropriate. Is there any means to compare the performance on the whole testsuite by running test jobs with some extra parameters?

[the remark this answers is not in the document]
Comment (casella) [reply OMCompiler/Compiler/BackEnd/Tearing.mo]: > 
> 
> On a slightly unrelated note
> [What Color Is Your Jacobian?](https://pdfs.semanticscholar.org/ae4d/65769b6551a51d6fc6be2f021515bffa0798.pdf):
> is actually written by my cousin. I guess I should read it from cover to cover. 😊😊

Whoa. I thought this was a joke until a saw the actual paper 😄
